### PR TITLE
Fix bug to skip only collection files that are not the latest

### LINKED ISSUE #408: Validate 2.1.0-SNAPSHOT skips a collection XML label
## 🐛 Describe the bug
The Validate tool is reporting that it's skipping collection_data_table.xml "due to collection not latest or does not sharing the same logical_identifier as the bundle target." The LID is the same in the collection product and the bundle, so the problem must be “collection not latest”. Since the collection product is listed in the bundle by its LID, not LIDVID, the rule is to assume the latest version of the collection is intended. So what makes Validate think this version is not the latest? To complicate matters, a few lines down the Validate tool reports that collection_data_table.xml passed. 

## 📜 To Reproduce
The bundle in question is the InSight SEIS bundle at https://pds-geosciences.wustl.edu/insight/urn-nasa-pds-insight_seis/data/. The bundle is too large to send a copy, so I'm attaching the Validate report, bundle.xml, and collection product.

## 🕵️ Expected behavior
I expect Validate not to skip the collection.

## 📚 Version of Software Used
Validate 2.1.0-SNAPSHOT

## 🩺 Test Data / Additional context
[validate.zip](https://github.com/NASA-PDS/validate/files/7226301/validate.zip)

##  🏞Screenshots
n/a

## 🖥 System Info
OS: Windows 10

---

### PR DESCRIPTION
**Pull request by qchaupds**
1. Add test resources : src/test/resources/github408
2. Correct logic of getLatestVersion() function to handle multiple collections : src/main/java/gov/nasa/pds/tools/util/LidVid.java
3. Add debug printing : src/main/java/gov/nasa/pds/tools/util/ReferentialIntegrityUtil.java
4. Correct logic to find latest collection files, remove multiple reporting of skip file and check for null-ness and size of otherCollectionIdList variable to avoid null pointer exception : src/main/java/gov/nasa/pds/tools/validate/BundleManager.java
5. Add stack trace printing to help debugging when exception occurs : src/main/java/gov/nasa/pds/validate/ValidateLauncher.java
6. Add new test : src/test/resources/features/validate.feature

Refs:

https://github.com/nasa-pds/validate/issues/408 Validate 2.1.0-SNAPSHOT skips a collection XML label

Closes https://github.com/nasa-pds/validate/issues/408 Validate 2.1.0-SNAPSHOT skips a collection XML label

User had reported that a collection was skipped in error.  As it turns out, multiple collections may have different versions for the same
collection.  Also, there was a bug with the logic of sorting of the versions.  When sorting based on character, "9.0" is larger than "10.0", which is not correct.
The fix was to sort the version using number order by creating a separate float array of the versions.
Code is fixed now and should only skip collection files that are of older versions.  The latest collection
is defined as the collection with the largest version_id value.
User also reported an extra validation of a collection that was supposively skipped.  It should not happen now.

Tested in DEV:

Case 1:
Running validate against a valid bundle with multiple collections with two collections with multiple versions.
The size of the test files (.csv) have been reduced to just a few records to reduce the size of the repository.
Also note that the content of the bundle in github408 resource is not the same as the URL reported in the ticket.  The
bundle contains just a few collections with smaller data files to allow the testing of multiple collections.

```
validate -R pds4.bundle --skip-content-validation -r report_github408_bundle_valid.json  -s json -t src/test/resources/github408/valid/bundle_insight_seis.xml
```

There should be 2 SKIP messages for 2 labels because the "extra" collection files have lower version number than the original collection:

```
{pds-dev3.jpl.nasa.gov}/home/qchau/sandbox/validate 138 % grep -A 4 SKIP report_github408_bundle_valid.json
      "status": "SKIP",
      "label": "file:/home/qchau/sandbox/validate/src/test/resources/github408/valid/data/collection_data_seed_2.xml",
      "messages": [
        {
          "severity": "INFO",
--
      "status": "SKIP",
      "label": "file:/home/qchau/sandbox/validate/src/test/resources/github408/valid/data/collection_data_table_2.xml",
      "messages": [
        {
          "severity": "INFO",
```

The collection files that ends with "_2.xml" contains a smaller version than the ones without, which should now be skipped by validate.

```
{pds-dev3.jpl.nasa.gov}/home/qchau/sandbox/validate 111 % egrep -A 1 "logical_ide" src/test/resources/github408/valid/data/*collection*.xml
src/test/resources/github408/valid/data/collection_data_lander.xml:        <logical_identifier>urn:nasa:pds:insight_seis:data_laf</logical_identifier>
src/test/resources/github408/valid/data/collection_data_lander.xml-        <version_id>4.0</version_id>
--
src/test/resources/github408/valid/data/collection_data_seed_2.xml:        <logical_identifier>urn:nasa:pds:insight_seis:data_seed</logical_identifier>
src/test/resources/github408/valid/data/collection_data_seed_2.xml-        <version_id>9.0</version_id>
--
src/test/resources/github408/valid/data/collection_data_seed.xml:        <logical_identifier>urn:nasa:pds:insight_seis:data_seed</logical_identifier>
src/test/resources/github408/valid/data/collection_data_seed.xml-        <version_id>10.0</version_id>
--
src/test/resources/github408/valid/data/collection_data_table_2.xml:        <logical_identifier>urn:nasa:pds:insight_seis:data_table</logical_identifier>
src/test/resources/github408/valid/data/collection_data_table_2.xml-        <version_id>9.0</version_id>
--
src/test/resources/github408/valid/data/collection_data_table.xml:        <logical_identifier>urn:nasa:pds:insight_seis:data_table</logical_identifier>
src/test/resources/github408/valid/data/collection_data_table.xml-        <version_id>10.0</version_id>
```

### CODE REVIEW

File: src/main/java/gov/nasa/pds/tools/util/LidVid.java [hunk 1]
```diff
@@ -43,21 +43,32 @@
...
-      String largestVersion = null;
-      if (vidsList.size() == 0) {
-          return(largestVersion);
+    HashMap<String,String> largestVersion = new HashMap<String,String>(); // A HashMap array containing a map of logical identifier and its latest version.
+
+    LOG.debug("getLatestVersion:lidvidsMap.size {}",lidvidsMap.size());
+
+    // Loop through each logical_identifier in lidvidsMap and find the largest version for that logical_identifier.
+    for (String lid : lidvidsMap.keySet()) {
+      ArrayList<String> vidsList = lidvidsMap.get(lid);  // The value in HashMap is an array of version_id list.
```
[the remark this answers is not in the document]
Comment (qchaupds) [reply src/main/java/gov/nasa/pds/tools/util/LidVid.java:59]: @sonatype-lift ignore

File: src/main/java/gov/nasa/pds/tools/util/LidVid.java [hunk 2]
```diff
@@ -127,32 +152,60 @@ public static String getLatestVersion(List<String> vidsList) {
...
       // The target with the largest version is the desired one to return.
       // Because largestVersion can be null if the list of targets is empty, it must be checked.
-      if (largestVersion != null) {
-          reducedKids.add(vidsTable.get(largestVersion));
-          LOG.debug("reduceToLatestTargetOnly:vidsTable.get(largestVersion) {}",vidsTable.get(largestVersion));
+      if (largestVersion != null && largestVersion.size() > 0) {
+          String fullyQualifiedLogicalIdentifierVersion = null;  // Also known as lidvid.
+          for (String singleLogicalIdentifier  : largestVersion.keySet()) {
+              // To build the key to vidsTable, it is necessary to add the logical identifier to "::" and then the version_id
+              fullyQualifiedLogicalIdentifierVersion = singleLogicalIdentifier + "::" + largestVersion.get(singleLogicalIdentifier);
```
[the remark this answers is not in the document]
Comment (qchaupds) [reply src/main/java/gov/nasa/pds/tools/util/LidVid.java:202]: @sonatype-lift ignore

File: src/main/java/gov/nasa/pds/tools/util/LidVid.java [hunk 3]
```diff
@@ -43,21 +43,32 @@
...
-      if (vidsList.size() == 0) {
-          return(largestVersion);
+    HashMap<String,String> largestVersion = new HashMap<String,String>(); // A HashMap array containing a map of logical identifier and its latest version.
+
+    LOG.debug("getLatestVersion:lidvidsMap.size {}",lidvidsMap.size());
+
+    // Loop through each logical_identifier in lidvidsMap and find the largest version for that logical_identifier.
+    for (String lid : lidvidsMap.keySet()) {
+      ArrayList<String> vidsList = lidvidsMap.get(lid);  // The value in HashMap is an array of version_id list.
+      LOG.debug("getLatestVersion:vidsList,vidsList.size {},{}",vidsList,vidsList.size());
```
[the remark this answers is not in the document]
Comment (qchaupds) [reply src/main/java/gov/nasa/pds/tools/util/LidVid.java:60]: @sonatype-lift ignore

**Review by jordanpadams — APPROVED** [review]

**Comment by jordanpadams** [comment]
Looks good.

**Pull request merged**